store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java

```diff
@@ -147,11 +150,139 @@ public void addServerProperties(VolumeInfo volInfo) throws ServiceException, JSO
 
             // step 6: Set ldap attribute
             provisioning.getLocalServer().setServerExternalStoreConfig(currentJsonObject.toString());
+
+            //If Unified volume is enable then go ahead to edit GlobalExternalStore
+            if (volInfo.getVolumeExternalInfo() != null
+                    && AdminConstants.A_VOLUME_S3.equalsIgnoreCase(volInfo.getVolumeExternalInfo().getStorageType())
+                    && volInfo.getVolumeExternalInfo().isUnified()) {
+                editGlobalExternalSettingsWhenAddVolume(volInfo, serverId);
+            }
+
         } catch (JSONException e) {
             throw e;
         }
     }
 
+    /**
+     * Modify zimbraGlobalExternalStoreConfig Ldap attribute,
+     * adds/update unified volumes
+     * @param volInfo
+     * @param serverId
+     * @throws JSONException, ServiceException
+     */
+    public void editGlobalExternalSettingsWhenAddVolume(VolumeInfo volInfo, String serverId) throws ServiceException, JSONException {
+
+        try {
+            String gescLdapJson = Provisioning.getInstance().getConfig().getGlobalExternalStoreConfig();
+
+            if (!StringUtil.isNullOrEmpty(gescLdapJson)) {
+                JSONObject gescLdapJO = new JSONObject(gescLdapJson);
+                JSONArray unifiedJsonObject = (JSONArray) gescLdapJO.opt(GlobalExternalStoreConfigConstants.A_S3_UNIFIED_VOLUME);
+
+                JSONArray volumeList = null;
+                JSONArray serverList = new JSONArray();
+                JSONObject server = null;
+
+                JSONObject newVolume = new JSONObject();
+                newVolume.put(GlobalExternalStoreConfigConstants.A_S3_VOLUME_ID, volInfo.getId());
+                newVolume.put(GlobalExternalStoreConfigConstants.A_S3_VOLUME_PREFIX, volInfo.getVolumeExternalInfo().getVolumePrefix());
+                newVolume.put(GlobalExternalStoreConfigConstants.A_S3_GLOBAL_BUCKET_CONFIG_ID,
+                        volInfo.getVolumeExternalInfo().getGlobalBucketConfigurationId());
+
```

Comment:
Before further processing check whether s3 buckets are created or not.. if not then may be we can fail this.

Reply:
Done